cmd/kubeadm/app/util/staticpod/utils.go

```diff
@@ -215,6 +214,32 @@ func ReadStaticPodFromDisk(manifestPath string) (*v1.Pod, error) {
 	return pod, nil
 }
 
+// GetEtcdProbeURL will return a validated --advertise-client-url URL or https://127.0.0.1:2379
+// to use for etcd liveness probes. in static pod manifests.
+func GetEtcdProbeURL(cfg *kubeadmapi.MasterConfiguration, componentName string) string {
+	if cfg.Etcd.ExtraArgs != nil {
+		if arg, exists := cfg.Etcd.ExtraArgs[etcdAdvertiseClientURLsArg]; exists {
+			// basic etcd url validation. Reference:
+			// https://github.com/coreos/etcd/blob/master/pkg/types/urls.go#L35-L47
+			u, err := url.Parse(arg)
```

Comment:
This `url.Parse` needs an error handler

given:
```golang
tests := []string{
        "http://[2001:db8::1]:2379",
        "http://domain.com:2379",
        "domain.com",
        "domain",
        "http://:2379",
        "http://2379",
        "hello://2379",
        ":2379",
        "http://",
        "://",
        "blah//",
        "",
}
```
It will fail with these errors and cause the code to panic
```
:2379
error parsing etcdAdvertiseClientURLsArg: parse :2379: missing protocol scheme

://
error parsing etcdAdvertiseClientURLsArg: parse ://: missing protocol scheme
```

I'd probably change it to:
```golang
if u, err := url.Parse(arg); err == nil {
  ...
}
```

Given that the function signature currently does not return an error, it's probably appropriate to return `https://127.0.0.1:2379` in the case of a Parse error.
This is the simplest thing to do, but may not be the best for the user.
It may be more appropriate to fail fatally, but this would likely require work to return an error and add error handling to the caller.

It may be an easy workaround to simply print a warning in the case that the flag fails to parse.
This is less of a big deal if there are already other functions that are implementing this workaround.

Reply:
It might be good to maintain support for `etcdListenClientURLsArg` as a fallback to `etcdAdvertiseClientURLsArg` so that we don't unnecessarily deprecate that functionality.
This would involve un-removing that code into this new function.

What do you think?

Reply:
( oh, it would be good to add tests for ipv6 urls as well as the invalid panic strings. )

Reply:
@stealthybox What do you think of something like this for the error handler part: https://play.golang.org/p/PEIQegS9fET